tests/unit/experimental/test_scope_parser.py

```diff
@@ -0,0 +1,321 @@
+    assert str(s) == "top[*subscope]"
+    assert repr(s) == "Scope('top', dependencies=[Scope('subscope', optional=True)])"
+
+
+def test_scope_parsing_allows_empty_string():
+    scopes = Scope.parse("")
+    assert scopes == []
+
+
+@pytest.mark.parametrize(
+    "scope_string1,scope_string2",
+    [
+        ("foo ", "foo"),
+        (" foo", "foo"),
+        ("foo[ bar]", "foo[bar]"),
+    ],
+)
+def test_scope_parsing_ignores_non_semantic_whitespace(scope_string1, scope_string2):
+    list1 = Scope.parse(scope_string1)
+    list2 = Scope.parse(scope_string2)
+    assert len(list1) == len(list2) == 1
+    s1, s2 = list1[0], list2[0]
+    # Scope.__eq__ is not defined, so equivalence checking is manual (and somewhat error
+    # prone) for now
+    assert s1.scope_string == s2.scope_string
+    assert s1.optional == s2.optional
+    for i in range(len(s1.dependencies)):
+        assert s1.dependencies[i].scope_string == s2.dependencies[i].scope_string
+        assert s1.dependencies[i].optional == s2.dependencies[i].optional
+
+
+@pytest.mark.parametrize(
+    "scopestring",
+    [
+        # ending in '*'
+        "foo*",
+        "foo *",
+        # '*' followed by '[] '
+        "foo*[bar]",
+        "foo *[bar]",
+        "foo [bar*]",
+        "foo * ",
+        "* foo",
+        # empty brackets
+        "foo[]",
+        # starting with open bracket
+        "[foo]",
+        # double brackets
+        "foo[[bar]]",
+        # unbalanced open brackets
+        "foo[",
+        "foo[bar",
+        # unbalanced close brackets
+        "foo]",
+        "foo bar]",
+        "foo[bar]]",
+        "foo[bar] baz]",
+        # space before brackets
+        "foo [bar]",
+        # missing space before next scope string after ']'
+        "foo[bar]baz",
+    ],
+)
+def test_scope_parsing_rejects_bad_inputs(scopestring):
+    with pytest.raises(ScopeParseError):
+        Scope.parse(scopestring)
+
+
+@pytest.mark.parametrize(
+    "scopestring",
+    [
+        "foo[foo]",
+        "foo[*foo]",
+        "foo[bar[foo]]",
+        "foo[bar[baz[bar]]]",
+        "foo[bar[*baz[bar]]]",
+        "foo[bar[*baz[*bar]]]",
+    ],
+)
+def test_scope_parsing_catches_and_rejects_cycles(scopestring):
+    with pytest.raises(ScopeCycleError):
+        Scope.parse(scopestring)
+
+
+def test_scope_parsing_catches_and_rejects_very_large_cycles_quickly():
+    """
+    WARNING: this test is hardware speed dependent and could fail on slow systems.
+
+    This test creates a very long cycle and validates that it can be caught in a
+    small timeframe of < 100ms.
+    Observed times on a test system were <20ms, and in CI were <60ms.
+
+    Although checking the speed in this way is not ideal, we want to avoid high
+    time-complexity in the cycle detection. This test offers good protection against any
+    major performance regression.
+    """
+    scope_string = ""
+    for i in range(1000):
+        scope_string += f"foo{i}["
+    scope_string += " foo10"
```

Comment:
Is it significant that there's a space, or that this points backwards to `foo10` instead of some other number?

Reply:
I think the space is a typo. Either that, or I wrote this test in some way that needed it (maybe generating `[foo{i}`?) and then refactored. It doesn't do anything, so I'll remove it.

`foo10` is pretty arbitrary, I just wanted a long chain that looked like this:
```
a -> b -> c -> d -> e -> f -> g -> ... -> z
     ^                                   /
      \                                 /
       `-------------------------------'
```

I'll add a comment to note that it's an arbitrary choice of where to introduce the cycle.